### PR TITLE
Using jsonpatch and podpreset to accept cluster config

### PR DESCRIPTION
**Pull request by TairuiWang**
all tests passed on internal cluster

### CODE REVIEW

**Comment by TairuiWang** [comment]
@aurickq ready for review

File: helm/adaptdl-sched/values.yaml [hunk 1]
```diff
@@ -33,6 +33,12 @@ job:
 #    limits:
 #      cpu: "1.0"
 #      memory: "1Gi"
+  # Users can add customized patch for pods created by adaptdl.
+  # jobPatch:
+  #   # podPatch will be applied to pod level
+  #   podPatch: "[{\"op\": \"add\", \"path\": \"/metadata/annotations/k8s.v1.cni.cncf.io~1networks\", \"value\": \"macvlan-conf\"}]"
+  #   # containerPatch will be applied to each container
+  #   containerPatch: "[{\"op\": \"add\", \"path\": \"/env/0\", \"value\": {\"name\": \"NCCL_SOCKET_IFNAME\", \"value\": \"net1\"}}]"
```
Comment (aurickq) [inline helm/adaptdl-sched/values.yaml:41]: I think the example could just be written in-line as YAML, which makes writing this section a bit more readable, like
```
- op: add
  path: "/metadata/annotations/k8s.v1.cni.cncf.io~1networks"
  value: "macvlan-conf"
```

File: helm/adaptdl-sched/templates/config.yaml [hunk 2]
```diff
@@ -10,3 +10,6 @@ data:
   {{- if .Values.job.defaultResources }}
   ADAPTDL_JOB_DEFAULT_RESOURCES: {{ .Values.job.defaultResources | toJson | quote }}
   {{- end }}
+  {{- if .Values.job.jobPatch }}
+  ADAPTDL_JOB_POD_PATCH : {{ .Values.job.jobPatch | toJson | quote }}
+  {{- end }}
```
Comment (aurickq) [inline helm/adaptdl-sched/templates/config.yaml:15]: Split into two env vars: `ADAPTDL_JOB_PATCH_PODS` and `ADAPTDL_JOB_PATCH_CONTAINERS` referring to `.Values.job.patch.pods` and `.Values.job.patch.containers`, respectively.

File: sched/adaptdl_sched/controller.py [hunk 3]
```diff
@@ -401,9 +402,24 @@ async def _create_pod(self, job_metadata, pod_template,
                     "name": "NVIDIA_VISIBLE_DEVICES",
                     "value": "none",
                 })
+        pod = self._pod_update(pod)
         await self._core_api.create_namespaced_pod(
             job_metadata["namespace"], pod)
 
+    def _pod_update(self, pod):
```
Comment (aurickq) [inline sched/adaptdl_sched/controller.py:409]: Use a more descriptive name, like `_patch_pod_and_containers`

**Review by aurickq — CHANGES_REQUESTED** [review]

File: helm/adaptdl-sched/templates/config.yaml [hunk 4]
```diff
@@ -10,3 +10,11 @@ data:
   {{- if .Values.job.defaultResources }}
   ADAPTDL_JOB_DEFAULT_RESOURCES: {{ .Values.job.defaultResources | toJson | quote }}
   {{- end }}
+  {{- if .Values.job.patch }}
+  {{- if .Values.job.patch.pods }}
+  ADAPTDL_JOB_POD_PATCH : {{ .Values.job.patch.pods | toJson | quote }}
```
Comment (aurickq) [inline helm/adaptdl-sched/templates/config.yaml:15]: `ADAPTDL_JOB_PATCH_PODS`

File: sched/adaptdl_sched/config.py [hunk 5]
```diff
@@ -57,3 +57,13 @@ def get_storage_subpath():
...
     val = os.getenv("ADAPTDL_JOB_DEFAULT_RESOURCES")
     return json.loads(val) if val is not None else None
+
+
+def get_job_pod_patch():
+    val = os.getenv("ADAPTDL_JOB_POD_PATCH")
+    return json.loads(val) if val is not None else None
+
+
+def get_job_container_patch():
```
Comment (aurickq) [inline sched/adaptdl_sched/config.py:67]: Keep getter names matching env var names: `get_job_patch_pods()` and `get_job_patch_containers()`

**Review by aurickq — CHANGES_REQUESTED** [review]

**Pull request merged**